internal/storage/inmem/store.go

```diff
@@ -0,0 +1,328 @@
+package inmem
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-memdb"
+
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Store implements an in-memory resource database using go-memdb.
+//
+// It can be used as a storage backend directly via the Backend type in this
+// package, but also handles reads in our Raft backend, and can be used as a
+// local cache when storing data in external systems (e.g. RDBMS, K/V stores).
+type Store struct {
+	db  *memdb.MemDB
+	pub *stream.EventPublisher
+}
+
+// NewStore creates a Store.
+//
+// You must call Run before using the store.
+func NewStore() (*Store, error) {
+	db, err := memdb.NewMemDB(&memdb.DBSchema{
+		Tables: map[string]*memdb.TableSchema{
+			tableNameMetadata: {
+				Name: tableNameMetadata,
+				Indexes: map[string]*memdb.IndexSchema{
+					indexNameID: {
+						Name:         indexNameID,
+						AllowMissing: false,
+						Unique:       true,
+						Indexer:      &memdb.StringFieldIndex{Field: "Key"},
+					},
+				},
+			},
+			tableNameResources: {
+				Name: tableNameResources,
+				Indexes: map[string]*memdb.IndexSchema{
+					indexNameID: {
+						Name:         indexNameID,
+						AllowMissing: false,
+						Unique:       true,
+						Indexer:      idIndexer{},
+					},
+					indexNameOwner: {
+						Name:         indexNameOwner,
+						AllowMissing: true,
+						Unique:       false,
+						Indexer:      ownerIndexer{},
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Store{
+		db:  db,
+		pub: stream.NewEventPublisher(10 * time.Second),
```

Comment:
❓ How do we arrive at this frequency?  Is this what is used in the server agent today? If so, may be adding comment for context would be helpful.

Reply:
Copied from here:

https://github.com/hashicorp/consul/blob/d5677e5680050a9b22403c160f5edde59613a4fc/agent/setup.go#L182

No idea how it was originally chosen 😄

Reply:
I don't know either but it seems reasonable. 

The snapshot points to the event buffer head at the time of its creation. As time moves forward new events get appended to the buffer and any active streams will go ahead and read those events from the stream. Even once all active streams have processed the events, the events are prevented from being garbage collected due to the snapshot pointing at them. 

Setting the snapshot cache ttl to 10s allows us to not be constantly recalculating snapshots for every new streaming request but isn't so high as to cause memory usage to get too high.